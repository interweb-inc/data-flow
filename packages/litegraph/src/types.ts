
export interface Litegraph {
  nodes: LitegraphNode[];
  links: LitegraphLink[];
}

export interface LitegraphNode {
  id: number;
  type: string;
  title: string;
  pos: [number, number]
  value?: any;
  outputs?: LitegraphNodeOutput[];
  inputs?: LitegraphNodeInput[];
}

export interface LitegraphNodeOutput {
  name: string;
  type: string;
  links: number[];
}

export interface LitegraphNodeInput {
  name: string;
  type: string;
  link: number | null;
}

export interface LitegraphLink {
  id: number;
  origin_id: number;
  origin_slot: number;
  target_id: number;
  target_slot: number;
}
