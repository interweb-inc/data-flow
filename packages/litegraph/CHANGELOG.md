# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.2.0](https://github.com/interweb-inc/data-flow/compare/@fbp/litegraph@0.1.0...@fbp/litegraph@0.2.0) (2024-05-28)

**Note:** Version bump only for package @fbp/litegraph





# 0.1.0 (2024-05-28)

**Note:** Version bump only for package @fbp/litegraph
