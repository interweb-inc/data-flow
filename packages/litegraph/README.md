# @fbp/litegraph

<p align="center" width="100%">
  <img src="https://github.com/cosmology-tech/interweb-utils/assets/545047/89c743c4-be88-409f-9a77-4b02cd7fe9a4" width="80">
  <br/>
  Flow-Based Programming schemas and converters for litegraph
  <br />
  <a href="https://github.com/interweb-inc/data-flow/actions/workflows/run-tests.yaml">
    <img height="20" src="https://github.com/interweb-inc/data-flow/actions/workflows/run-tests.yaml/badge.svg" />
  </a>
   <a href="https://github.com/interweb-inc/data-flow/blob/main/LICENSE">
    <img height="20" src="https://img.shields.io/badge/license-MIT-blue.svg"/>
  </a>
</p>

## install

```sh
npm install litegraph
```
