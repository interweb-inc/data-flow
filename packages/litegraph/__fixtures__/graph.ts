import { Graph } from '@fbp/types';

export const graph: Graph = {
  type: 'Graph',
  name: 'default',
  nodes: [
    {
      name: 'add1',
      type: 'js/math/add',
      inputs: [
        {
          name: 'A',
          type: 'number',
        },
        {
          name: 'B',
          type: 'number',
        },
      ],
      meta: {
        x: 100,
        y: 100,
      },
      outputs: [
        {
          name: 'Sum',
          type: 'number',
        },
      ],
      properties: [],
    },
    {
      name: 'mult1',
      type: 'js/math/mult',
      inputs: [
        {
          name: 'Sum',
          type: 'number',
        },
        {
          name: 'C',
          type: 'number',
        },
      ],
      meta: {
        x: 300,
        y: 100,
      },
      outputs: [
        {
          name: 'Product',
          type: 'number',
        },
      ],
      properties: [],
    }
  ],
  edges: [
    {
      src: {
        node: 'add1',
        port: 'Sum'
      },
      dst: {
        node: 'mult1',
        port: 'Sum'
      }
    }
  ]
};
