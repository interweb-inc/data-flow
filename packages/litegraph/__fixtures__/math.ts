import { Graph } from '@fbp/types';

export const graph: Graph = {
  type: 'Graph',
  name: 'default',
  nodes: [
    {
      name: 'num1',
      type: 'js/const/number',
      meta: {
        x: 50,
        y: 200,
      },
      outputs: [
        {
          name: 'num',
          type: 'number',
        },
      ],
      properties: [],
    },
    {
      name: 'num2',
      type: 'js/const/number',
      meta: {
        x: 350,
        y: 200,
      },
      outputs: [
        {
          name: 'num',
          type: 'number',
        },
      ],
      properties: [],
    },
    {
      name: 'add1',
      type: 'js/math/add',
      inputs: [
        {
          name: 'a',
          type: 'number',
        },
        {
          name: 'b',
          type: 'number',
        },
      ],
      meta: {
        x: 100,
        y: 100,
      },
      outputs: [
        {
          name: 'sum',
          type: 'number',
        },
      ],
      properties: [],
    },
    {
      name: 'mult1',
      type: 'js/math/mult',
      inputs: [
        {
          name: 'a',
          type: 'number',
        },
        {
          name: 'b',
          type: 'number',
        },
      ],
      meta: {
        x: 300,
        y: 100,
      },
      outputs: [
        {
          name: 'product',
          type: 'number',
        },
      ],
      properties: [],
    }
  ],
  edges: [
    {
      src: {
        node: 'num1',
        port: 'num'
      },
      dst: {
        node: 'add1',
        port: 'b'
      }
    },
    {
      src: {
        node: 'num2',
        port: 'num'
      },
      dst: {
        node: 'add1',
        port: 'b'
      }
    },
    {
      src: {
        node: 'add1',
        port: 'sum'
      },
      dst: {
        node: 'mult1',
        port: 'a'
      }
    },
    {
      src: {
        node: 'num1',
        port: 'num'
      },
      dst: {
        node: 'mult1',
        port: 'b'
      }
    }
  ]
};