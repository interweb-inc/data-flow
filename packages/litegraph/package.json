{
  "name": "@fbp/litegraph",
  "version": "0.2.0",
  "author": "Dan Lynch <pyramation@gmail.com>",
  "description": "litegraph",
  "main": "index.js",
  "module": "esm/index.js",
  "types": "index.d.ts",
  "homepage": "https://github.com/interweb-inc/data-flow",
  "license": "SEE LICENSE IN LICENSE",
  "publishConfig": {
    "access": "public",
    "directory": "dist"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/interweb-inc/data-flow"
  },
  "bugs": {
    "url": "https://github.com/interweb-inc/data-flow/issues"
  },
  "scripts": {
    "copy": "copyfiles -f ../../LICENSE README.md package.json dist",
    "clean": "rimraf dist/**",
    "prepare": "npm run build",
    "build": "npm run clean; tsc; tsc -p tsconfig.esm.json; npm run copy",
    "lint": "eslint . --fix",
    "test": "jest",
    "test:watch": "jest --watch"
  },
  "dependencies": {
    "@fbp/types": "^0.2.0",
    "litegraph.js": "^0.7.18"
  },
  "keywords": []
}
