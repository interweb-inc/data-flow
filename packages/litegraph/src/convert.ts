import { Edge, Graph, Node } from '@fbp/types';

import { Litegraph, LitegraphLink, LitegraphNode } from './types';

export function convertToLitegraph(graph: Graph): Litegraph {
  const nodes: LitegraphNode[] = graph.nodes.map((node, index) => ({
    id: index,  // Assign an incremental ID based on the array index
    type: [node.context, node.category, node.type].join('/'),
    title: node.name,
    // Ensure pos is always a tuple of [number, number]
    pos: [node.meta?.x || 0, node.meta?.y || 0] as [number, number],
    value: node.properties?.reduce((acc, prop) => ({ ...acc, [prop.name]: prop.value }), undefined),
    inputs: node.inputs.map(input => ({
      name: input.name,
      type: input.type,
      link: null  // Initial null, to be populated in the links processing
    })),
    outputs: node.outputs.map(output => ({
      name: output.name,
      type: output.type,
      links: []  // Initial empty array, to be populated in the links processing
    }))
  }));

  const links: LitegraphLink[] = [];

  // Process each node to find connections
  graph.edges.forEach((edge, edgeIndex) => {
    const srcNodeIndex = nodes.findIndex(n => n.title === edge.src.node);
    const dstNodeIndex = nodes.findIndex(n => n.title === edge.dst.node);

    if (srcNodeIndex !== -1 && dstNodeIndex !== -1) {
      const srcOutputIndex = nodes[srcNodeIndex].outputs.findIndex(output => output.name === edge.src.port);
      const dstInputIndex = nodes[dstNodeIndex].inputs.findIndex(input => input.name === edge.dst.port);

      // Create link
      const link: LitegraphLink = {
        id: edgeIndex,
        origin_id: srcNodeIndex,
        origin_slot: srcOutputIndex,
        target_id: dstNodeIndex,
        target_slot: dstInputIndex
      };
      links.push(link);

      // Update nodes with link information
      nodes[srcNodeIndex].outputs[srcOutputIndex].links.push(edgeIndex);
      nodes[dstNodeIndex].inputs[dstInputIndex].link = edgeIndex;
    }
  });

  return { nodes, links };
}


export function convertFromLitegraph(litegraph: Litegraph): Graph {
  const nodes: Node[] = litegraph.nodes.map(node => {
    const [context,category,type] = node.type.split('/');
    const lightNode: Node = {
      name: node.title,
      context,
      category,
      type,
      meta: {
        x: node.pos[0],
        y: node.pos[1]
      },
      inputs: node.inputs?.map(input => ({
        name: input.name,
        type: input.type
      })) || [],
      outputs: node.outputs?.map(output => ({
        name: output.name,
        type: output.type
      })) || [],
      properties: node.value ? Object.keys(node.value).map(key => ({
        name: key,
        type: typeof node.value[key], // Assumes all values are singular types
        value: node.value[key]
      })) : []
    };
    return lightNode;
  });

  const edges: Edge[] = litegraph.links.map(link => ({
    src: {
      node: litegraph.nodes[link.origin_id].title,
      port: litegraph.nodes[link.origin_id].outputs[link.origin_slot].name
    },
    dst: {
      node: litegraph.nodes[link.target_id].title,
      port: litegraph.nodes[link.target_id].inputs[link.target_slot].name
    }
  }));

  return {
    category: 'default',
    context: 'default',
    name: 'default',
    type: 'Graph',
    nodes,
    edges
  };
}
