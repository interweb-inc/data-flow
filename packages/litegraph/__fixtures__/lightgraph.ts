import { Litegraph } from '../src';

export const lightgraph: Litegraph = {
  links: [],
  nodes: [
    {
      id: 0,
      inputs: [
        {
          link: null,
          name: 'A',
          type: 'number',
        },
        {
          link: null,
          name: 'B',
          type: 'number',
        },
      ],
      outputs: [
        {
          links: [],
          name: 'Sum',
          type: 'number',
        },
      ],
      pos: [
        100,
        100,
      ],
      title: 'Adder Node',
      type: 'Add',
      value: undefined,
    },
    {
      id: 1,
      inputs: [
        {
          link: null,
          name: 'Sum',
          type: 'number',
        },
        {
          link: null,
          name: 'C',
          type: 'number',
        },
      ],
      outputs: [
        {
          links: [],
          name: 'Product',
          type: 'number',
        },
      ],
      pos: [
        300,
        100,
      ],
      title: 'Multiplier Node',
      type: 'Multiply',
      value: undefined,
    },
  ],
};