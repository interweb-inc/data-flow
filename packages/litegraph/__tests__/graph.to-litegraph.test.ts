import { graph as simple} from '../__fixtures__/graph';
import { litegraph } from '../__fixtures__/litegraph';
import { graph } from '../__fixtures__/math';
import { convertToLitegraph } from '../src/convert';

it('convertToLitegraph', () => {
  const result = convertToLitegraph(simple);
  expect(result).toEqual(litegraph);
});

it('convertToLitegraph', () => {
  const result = convertToLitegraph(graph);
  expect(result).toMatchSnapshot();
});