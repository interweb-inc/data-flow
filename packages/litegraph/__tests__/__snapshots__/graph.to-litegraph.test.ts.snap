// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertToLitegraph 1`] = `
{
  "links": [
    {
      "id": 0,
      "origin_id": 0,
      "origin_slot": 0,
      "target_id": 2,
      "target_slot": 1,
    },
    {
      "id": 1,
      "origin_id": 1,
      "origin_slot": 0,
      "target_id": 2,
      "target_slot": 1,
    },
    {
      "id": 2,
      "origin_id": 2,
      "origin_slot": 0,
      "target_id": 3,
      "target_slot": 0,
    },
    {
      "id": 3,
      "origin_id": 0,
      "origin_slot": 0,
      "target_id": 3,
      "target_slot": 1,
    },
  ],
  "nodes": [
    {
      "id": 0,
      "inputs": [],
      "outputs": [
        {
          "links": [
            0,
            3,
          ],
          "name": "num",
          "type": "number",
        },
      ],
      "pos": [
        50,
        200,
      ],
      "title": "num1",
      "type": "js/const/number",
      "value": undefined,
    },
    {
      "id": 1,
      "inputs": [],
      "outputs": [
        {
          "links": [
            1,
          ],
          "name": "num",
          "type": "number",
        },
      ],
      "pos": [
        350,
        200,
      ],
      "title": "num2",
      "type": "js/const/number",
      "value": undefined,
    },
    {
      "id": 2,
      "inputs": [
        {
          "link": null,
          "name": "a",
          "type": "number",
        },
        {
          "link": 1,
          "name": "b",
          "type": "number",
        },
      ],
      "outputs": [
        {
          "links": [
            2,
          ],
          "name": "sum",
          "type": "number",
        },
      ],
      "pos": [
        100,
        100,
      ],
      "title": "add1",
      "type": "js/math/add",
      "value": undefined,
    },
    {
      "id": 3,
      "inputs": [
        {
          "link": 2,
          "name": "a",
          "type": "number",
        },
        {
          "link": 3,
          "name": "b",
          "type": "number",
        },
      ],
      "outputs": [
        {
          "links": [],
          "name": "product",
          "type": "number",
        },
      ],
      "pos": [
        300,
        100,
      ],
      "title": "mult1",
      "type": "js/math/mult",
      "value": undefined,
    },
  ],
}
`;
